### PR TITLE
Fold compound expressions with literal arguments into literals

### PR DESCRIPTION
**Pull request by anandthakker**
When parsing an expression, if its arguments are all Literals, then evaluate it immediately and yield a Literal instead.

Yields a modest perf improvement in `tile_layout_dds`:
 - before: 51.37629 - 52.20937 ms / iteration
 - after: 47.81951 - 49.86712 ms / iteration

More importantly, this means we get validation errors at compile time rather than evaluation time for the common case of color curves defined like `[curve, [linear], [zoom], 0, "red", 10, "bleu mispelled"]`

benchmark | master 2200752 | expressions-fold-constant 0567a8f
--- | --- | ---
**map-load** | 111 ms  | 119 ms 
**style-load** | 128 ms  | 168 ms 
**buffer** | 1,167 ms  | 1,120 ms 
**tile_layout_dds** | 1,329 ms  | 1,255 ms 
**fps** | 60 fps  | 60 fps 
**frame-duration** | 7.1 ms, 1% > 16ms  | 6.9 ms, 2% > 16ms 
**query-point** | 0.98 ms  | 1.24 ms 
**query-box** | 73.43 ms  | 75.72 ms 
**geojson-setdata-small** | 17 ms  | 0 ms 
**geojson-setdata-large** | 88 ms  | 175 ms 
**filter** | Done  | Done

### CODE REVIEW

**Comment by anandthakker** [comment]
Example generated code for a color curve with `"red"` as the stop output value (excerpt of the `curve/linear-color` integration test output)

## before

```js
function e4() {   return "red" }
var v5 = [e4];
function e6() {   return $this.toColor(v5) }
```

## after

```js
var v4 = (new $this.Color(1, 0, 0, 1));
function e5() {   return v4 }
```

**Comment by mourner** [comment]
Very cool!

**Comment by anandthakker** [comment]
Rebased, generalized, and updated benchmarks.

File: src/style-spec/function/evaluation_context.js [hunk 1]
```diff
@@ -13,7 +13,6 @@ const {
     toString,
     checkSubtype} = require('./types');
 const {Color, typeOf} = require('./values');
-const Curve = require('./definitions/curve');
```
Comment (jfirebaugh) [inline src/style-spec/function/evaluation_context.js]: What's the reason for these changes?

File: src/style-spec/function/definitions/literal.js [hunk 2]
```diff
@@ -45,9 +45,20 @@ class Literal implements Expression {
...
+    compile(ctx: CompilationContext) {
+        let value;
+        if (this.type.kind === 'Color') {
+            value = `(new $this.Color(${(this.value: any).join(', ')}))`;
+        } else {
+            value = Literal.compile(this.value);
+        }
+
+        if (typeof this.value === 'object' && this.value !== null) {
+            const v = ctx.addVariable(value);
```
Comment (jfirebaugh) [inline src/style-spec/function/definitions/literal.js]: Nit: `return ctx.addVariable(value);`

File: src/style-spec/function/parse_expression.js [hunk 3]
```diff
@@ -68,4 +87,21 @@ function parseExpression(expr: mixed, context: ParsingContext): ?Expression {
     }
 }
 
+const nonConstantExpressions = [ 'error', 'get', 'has', 'properties', 'id', 'geometry-type', 'zoom' ];
```
Comment (jfirebaugh) [inline src/style-spec/function/parse_expression.js]: This looks similar to `isFeatureConstant`/`isZoomConstant`. Can they share more code?

File: src/style-spec/function/evaluation_context.js (hunk 1 above)
Comment (anandthakker) [reply src/style-spec/function/evaluation_context.js]: Breaking a new dependency cycle, which was introduced by `parse_expression` now depending on `evaluation_context`.  I guess it's better to break it there instead (especially since we're already doing so in that module)

File: src/style-spec/function/parse_expression.js (hunk 3 above)
Comment (anandthakker) [reply src/style-spec/function/parse_expression.js]: 👍 I almost did that, and then stopped myself because I thought I was prematurely abstracting.

Comment (anandthakker) [reply src/style-spec/function/parse_expression.js]: Oh, actually the code itself is a bit different. `isConstant` only looks one level deep (because that's sufficient for the way we're using it).  We could consolidate the base case code, though

**Comment by anandthakker** [comment]
@jfirebaugh updated to address your notes/questions
